### ISSUE TITLE
[Java] Map swagger "date" type to java.sql.Date

### ISSUE THREAD

**Issue by cbornet** [swagger-api/swagger-codegen#3733]
Currently, both "date-time" and "date" are mapped to java.util.Date if the date library is not joda nor java8.
It would be better to map "date" to ```java.sql.Date``` which is a subclass of java.util.Date that represents a date without time.
This will allow to ser/deser"date" and "date-time" in distinct formats.

**Comment by wing328** [comment]
@cbornet what about adding another option under `dateLibrary`?
```
	dateLibrary
	    Option. Date library to use
	        java8-localdatetime - Java 8 using LocalDateTime (for legacy app only)
	        joda - Joda
	        java8 - Java 8 native
	        legacy - Legacy java.util.Date
```

**Comment by cbornet** [comment]
@wing328 You mean for downward compatibility ? This is going to be quite messy !
I don't think that people currently using the "legacy" dateLibrary can correctly use the swagger "date" type. I've looked in all generators and they all serialize it as a "date-time". And only okhttp-gson seems to be able to deserialize it.
So it's more a fix for a completly broken feature than a breaking change.
But if you think that's a breaking change, I can create a "legacy-without-sqldate" lib.

**Comment by wing328** [comment]
@cbornet what about replacing `java.util.Date` with `java.sql.Date` starting 2.3.0? if later there're  demands to use `java.util.Date` with valid use cases, we add back `java.util.Date` as an option.

**Comment by cbornet** [comment]
OK. I'll PR against 2.3.0 branch

**Comment by brentryan** [comment]
@wing328 @cbornet I'm not sure I agree with changing to java.sql.Date as the default.  I can't think of a use case that would break off the top of my head, but I know from prior experience that using java.sql.Date has caused issues.  I think some of the issues (oddly enough) was around ORM libraries like hibernate/mybatis.  But I think using "java.sql.Date" seems wrong to me for models especially when many applications don't even use sql.

**Comment by cbornet** [comment]
@brentryan then what would be your solution for supporting the "date (rfc3339 full-date)" format without joda or java8 ?

**Comment by cbornet** [comment]
See #3648 for an example that our current usage of java.util.Date for ```date``` format is broken.

**Comment by cbornet** [comment]
And don't get fooled by the "sql" in java.sql.Date: it's just a subclass of java.util.Date with hour/minute/seconds set to 0. See https://docs.oracle.com/javase/7/docs/api/java/sql/Date.html

**Comment by brentryan** [comment]
This seems to work fine for me.  Maybe it's a serialization issue?
```

import com.fasterxml.jackson.databind.ObjectMapper;
import static com.fasterxml.jackson.databind.SerializationFeature.WRITE_DATES_AS_TIMESTAMPS;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.TimeZone;

public class DateTest {

    public static void main(String[] args) throws Exception {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
        sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
        ObjectMapper m = new ObjectMapper();
        Date d = new Date();
        m.setDateFormat(sdf);
        m.configure(WRITE_DATES_AS_TIMESTAMPS, false);
        
        Data data = new Data();
        data.date = d;
        
        String jsonOut = m.writeValueAsString(data);
        System.out.println("json: " + jsonOut);
        
        String dataJson = "{\"date\":\"2016-09-07T15:50:27.253Z\"}";
        Data parsedData = m.readValue(dataJson, Data.class);
        
        System.out.println("data: " + parsedData.date);
    }
    
    static class Data {
        public Date date;
    }
}
```

Or maybe I'm still missing what's broken here?

**Comment by cbornet** [comment]
Whay you are describing is a test for ```date-time``` format, not ```date```

**Comment by brentryan** [comment]
Do you have a simple test that demonstrates the issue?

**Comment by brentryan** [comment]
Is the issue that if I change the above to test to include just the date part of the format then the output winds up being: 
```
json: {"date":"2016-09-08"}
data: Wed Sep 07 20:00:00 EDT 2016
```

Basically it assumes the date is in EST on 9/8 when it gets output, but when it parses the string back in it doesn't know what timezone it's coming from so it assumes UTC and converts it to EDT. which makes the date 9/7.

Is this the issue?

**Comment by cbornet** [comment]
No, the issue is for instance with
```yaml
  SomeModel:
    type: object
    properties:
      someDate:
        type: string
        format: date
      someDateTime:
        type: string
        format: date-time
```
someDate must be (de)serialized with "yyyy-MM-dd" (full-date in rfc3339) and someDateTime with "yyyy-MM-dd'T'HH:mm:ss.SSSXXX" (date-time in rfc3339).

**Comment by cbornet** [comment]
For the change to be useful, the java.sql.Date must be correctly ser/der with format "yyyy-MM-dd".
* Jackson already correctly serializes but for deserialization, it uses the ObjectMapper DateFormat so we can't use SimpleDateFormat as we can't configure it for both date and date-time. We need to use ISO8601DateFormat instead that supports both formats. The good news is that it is the same need as #3727 
* Gson uses by default SimpleDateFormat("MMM d, yyyy") so it needs to be changed. For convenience, I have done it as part of #3790 

So once all PRs are merged and #3727 is closed, this is just a simple switch of typeMapping in AbstractJavaCodegen.
(Note that users nostalgic of the java.util.Date will still be able to override the default with --type-mappings option.)

**Comment by brentryan** [comment]
I'm still very much against this change.  It can cause lots of issues downstream with apis using jdbc.  This link provides lots of reasons around this and I know I've hit several of them in the past.  http://stackoverflow.com/questions/2305973/java-util-date-vs-java-sql-date

**Comment by cbornet** [comment]
The SOF you linked says that when using JDBC, you should store your dates as long or use joda or java8 dates. I can only agree with that. The java.util.Date and its subclass is flawed but java.sql.Date not more than java.util.Date.
This is for people who want to use the legacy dates anyway mostly on android to keep a low memory footprint.
We need different types for date and date-time and that's just what java.sql.Date represents.
If you have a better solution please explain it

**Comment by brentryan** [comment]
Yes, but the problem is that there's still lots of legacy out there that relies on java.util.Date that this wouldn't interopt with because those methods take a java.util.Date, not a java.sql.Date.  It's also recommended to pass a java.util.Date to the jdbc driver, not a java.sql.Date because of how the drivers interpret the right underlying database type.  java.util.Date might suck, but it still contains time information so I don't understand what the issue is.  It still seems to me like you're trying to solve a JSON serialization issue.  Can't you fix this by just setting your DateFormat properly and parsing it correctly?

**Comment by cbornet** [comment]
I'm starting to think that you're trolling me...
Please explain your solution for dealing with two different DateFormat for the same java type !

**Comment by brentryan** [comment]
lol. Trolling would be funny. I'm probably just missing something obvious. I honestly haven't done jdbc stuff in a while now. This just feels like a mistake to me. 

If I get time, I'll try to look at this a little closer this weekend. 

It sounds like you're trying to set 2 different DateFormat by using the Java Type to determine the difference. For us, we always just do ISO 8601 format everywhere in UTC and we let the presentation tier apply the date format as needed or we use a string data type and apply the custom date format rules.

**Comment by brentryan** [comment]
@cbornet Have you considered using jackson's @JsonFormat option which allows different formats per property?

```
Can I configure it on per-property basis?

Yes, with Jackson 2.0 you can use the new @JsonFormat annotation:

  public class DateStuff {
    @JsonFormat(shape=JsonFormat.Shape.STRING, pattern="yyyy-MM-dd,HH:00", timezone="CET")
    public Date creationTime;
  }
Notes on java.sql.Date

(aka "Please do NOT use java.sql.Date, ever!")

Although Jackson supports java.sql.Date, there are known issues with respect to timezone handling, partly due to design of this class. It is recommended that this type is avoided, if possible, and regular java.util.Date (or java.util.Calendar) used instead. If this is not possible, it may be necessary for applications to convert these dates using java.util.Calendar and explicit timezone definition.
```

Take from http://wiki.fasterxml.com/JacksonFAQDateHandling

**Comment by cbornet** [comment]
Yes, I already considered it. Apart from the complexity it adds in the model generation, it doesn't work with Gson.

**Comment by cbornet** [comment]
And yes, you should always wrap a java.sql.Date into a Calendar before using it (the getYear()/getMonth()/etc methods are deprecated).
And yes, you should avoid the whole java.util.Date/Calendar/etc... if you can and use joda.
